File: scripts/playable_package_serve.mjs
// Node fallback launcher (used only when python3 is unavailable) — zero
// dependencies, serves THIS package directory on 127.0.0.1.
//   node start-world-playable.mjs [port]
import { createServer } from 'node:http';
import { readFile, stat } from 'node:fs/promises';
import { extname, join, normalize, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = dirname(fileURLToPath(import.meta.url));
const types = { '.html': 'text/html; charset=utf-8', '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8', '.glb': 'model/gltf-binary', '.bin': 'application/octet-stream',
  '.png': 'image/png', '.jpg': 'image/jpeg', '.wasm': 'application/wasm', '.css': 'text/css; charset=utf-8' };

const port = Number(process.argv[2]) || 5411;
const server = createServer(async (req, res) => {
  try {
    let p = decodeURIComponent(new URL(req.url, 'http://x').pathname);
    let file = normalize(join(root, p));
    if (!file.startsWith(root)) { res.writeHead(403); res.end(); return; }
    const st = await stat(file).catch(() => null);
    if (st?.isDirectory()) file = join(file, 'index.html');
    const body = await readFile(file);
    res.writeHead(200, { 'content-type': types[extname(file)] ?? 'application/octet-stream', 'cache-control': 'no-store' });
    res.end(body);
  } catch {
    res.writeHead(404); res.end('not found');
  }
});
server.on('error', (e) => {
  console.error(`端口 ${port} 无法使用（${e.code}）。请换一个端口：node start-world-playable.mjs 5412`);
  process.exit(1);
});
server.listen(port, '127.0.0.1', () =>
  console.log(`方浜市声 本地试玩包已启动：http://127.0.0.1:${port}/world-preview.html （Ctrl+C 停止）`));

File: scripts/playable_package_serve.py
#!/usr/bin/env python3
# world-playable package launcher — python3 standard library only.
# Serves THIS directory (the package root) on 127.0.0.1 at the given port
# (default 5411). A busy port is reported clearly and the launcher exits
# non-zero; it never kills or waits on whatever holds the port.
#   python3 start-world-playable.py [port]
import os
import socket
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

ROOT = os.path.dirname(os.path.abspath(__file__))


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *a, **kw):
        super().__init__(*a, directory=ROOT, **kw)

    def end_headers(self):
        # local candidate only: no caching surprises, no cross-origin use
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, fmt, *args):  # quiet by default, errors still show
        if args and (" 4" in str(args[1][:3]) if len(args) > 1 and isinstance(args[1], int) else False):
            sys.stderr.write(" %s\n" % (fmt % args))


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 5411
    host = "127.0.0.1"
    try:
        srv = ThreadingHTTPServer((host, port), Handler)
    except OSError as e:
        print(f"端口 {port} 无法使用（{e}）。"
              f"请换一个端口：python3 start-world-playable.py 5412", file=sys.stderr)
        return 1
    print(f"方浜市声 本地试玩包已启动：http://{host}:{port}/world-preview.html"
          f" （Ctrl+C 停止；本服务只在 {host} 上监听）")
    try:
        srv.serve_forever()
    except KeyboardInterrupt:
        print("\n已停止。")
    return 0


if __name__ == "__main__":
    sys.exit(main())
